### PR TITLE
Hooking up sending

### PR DESCRIPTION
**Pull request by Bilb**
Hooking up the sending of our message to the new sending pipelie

### CODE REVIEW

File: libtextsecure/message_receiver.js [hunk 1]
```diff
@@ -1086,34 +1086,63 @@ MessageReceiver.prototype.extend({
...
+        );
+
+        await Promise.all([signedKeyPromise, preKeyPromise]);
+
         const device = {
           identityKey,
           deviceId,
-          preKey,
-          signedPreKey: signedKey,
+          preKey: {...preKey, keyId: preKeyId},
```
Comment (Mikunj) [inline libtextsecure/message_receiver.js:1144]:
```suggestion
          preKey: preKeyObject
```

File: ts/session/sending/MessageQueue.ts [hunk 2]
```diff
@@ -46,10 +46,12 @@ export class MessageQueue implements MessageQueueInterface {
     devices: Array<PubKey>,
     message: ContentMessage
   ) {
+    await this.pendingMessageCache.isReady;
+
     let currentDevices = [...devices];
 
     // Sync to our devices if syncable
-    if (SyncMessageUtils.canSync(message)) {
+    if (await SyncMessageUtils.canSync(message)) {
```
Comment (vincentbavitz) [inline ts/session/sending/MessageQueue.ts]: No need to await; this is now synchronous

File: libtextsecure/sendmessage.js [hunk 3]
```diff
@@ -550,57 +529,60 @@ MessageSender.prototype = {
...
-        contentMessage,
-        Date.now(),
-        silent,
-        options
-      );
+      const requestContactSyncMessage = new RequestContactSyncMessage({
+        timestamp: Date.now(),
+        reqestType: CONFIGURATION,
+      });
+      return libsession.getMessageQueue().send(user, requestContactSyncMessage);
```
Comment (Mikunj) [inline libtextsecure/sendmessage.js]:
```suggestion
      return libsession.getMessageQueue().sendSyncMessage( requestContactSyncMessage);
```
I know that in the future we should probably have a way to send a sync message to a specific device but for now this should be fine.

File: libtextsecure/sendmessage.js [hunk 4]
```diff
@@ -550,57 +529,60 @@ MessageSender.prototype = {
...
-        contentMessage,
-        Date.now(),
-        silent,
-        { ...options, debugMessageType }
-      );
+      const requestContactSyncMessage = new RequestContactSyncMessage({
+        timestamp: Date.now(),
+        reqestType: GROUPS,
+      });
+      return libsession.getMessageQueue().send(user, requestContactSyncMessage);
```
Comment (Mikunj) [inline libtextsecure/sendmessage.js]: same as above

File: ts/session/protocols/MultiDeviceProtocol.ts [hunk 5]
```diff
@@ -144,6 +148,11 @@ export class MultiDeviceProtocol {
   ): Promise<Array<PubKey>> {
     const pubKey = typeof user === 'string' ? new PubKey(user) : user;
     const authorisations = await this.getPairingAuthorisations(pubKey);
+    if (authorisations.length === 0) {
+      const array: Array<PubKey> = new Array();
```
Comment (Mikunj) [inline ts/session/protocols/MultiDeviceProtocol.ts:152]: does this not work?
```
return [pubKey];
```

Comment (Mikunj) [reply ts/session/protocols/MultiDeviceProtocol.ts:152]: We should also add a test case for this

File: ts/session/protocols/MultiDeviceProtocol.ts [hunk 6]
```diff
@@ -84,6 +84,10 @@ export class MultiDeviceProtocol {
     );
     // TODO: Filter out invalid authorisations
 
+    if (!mapping || !mapping.authorisations) {
```
Comment (Mikunj) [inline ts/session/protocols/MultiDeviceProtocol.ts:87]: need to add test case for this too

File: js/models/conversations.js [hunk 7]
```diff
@@ -389,21 +389,42 @@
...
+        const device = new libsession.Types.PubKey(recipientId);
+        libsession
+          .getMessageQueue()
+          .sendUsingMultiDevice(device, typingMessage)
+          .ignore();
+      } else {
+        // the recipients on the case of a group are found by the messageQueue using message.groupId
+        libsession
+          .getMessageQueue()
+          .sendToGroup(typingMessage)
```
Comment (Mikunj) [inline js/models/conversations.js:425]: We need to handle this case in `sendToGroup`

File: js/models/conversations.js [hunk 8]
```diff
@@ -1681,28 +1728,28 @@
...
-        promise = textsecure.messaging.sendExpirationTimerUpdateToGroup(
-          this.get('id'),
-          this.getRecipients(),
-          expireTimer,
-          message.get('sent_at'),
-          profileKey,
-          sendOptions
-        );
-      }
+        const expirationTimerMessage = new new libsession.Messages.Outgoing.ExpirationTimerUpdateMessage(
```
Comment (Mikunj) [inline js/models/conversations.js:1738]:
```suggestion
        const expirationTimerMessage = new libsession.Messages.Outgoing.ExpirationTimerUpdateMessage(
```

File: js/models/conversations.js [hunk 9]
```diff
@@ -1681,28 +1728,28 @@
...
+          expireUpdate
+        )();
 
-      await message.send(this.wrapSend(promise));
+        await libsession
+          .getMessageQueue()
+          .sendUsingMultiDevice(this.get('id'), expirationTimerMessage);
```
Comment (Mikunj) [inline js/models/conversations.js:1744]: `sendUsingMultiDevice` expects a `PubKey`

File: js/models/conversations.js [hunk 10]
```diff
@@ -1681,28 +1728,28 @@
...
+      } else {
+        expireUpdate.groupId = this.get('id');
+        const expirationTimerMessage = new new libsession.Messages.Outgoing.ExpirationTimerUpdateMessage(
+          expireUpdate
+        )();
 
+        await libsession.getMessageQueue().sendToGroup(expirationTimerMessage);
```
Comment (Mikunj) [inline js/models/conversations.js:1751]: this case will also need to be added in `sendToGroup`

File: js/models/conversations.js [hunk 11]
```diff
@@ -1868,33 +1911,53 @@
...
+        name: this.get('name'),
+        avatar: this.get('avatar'),
+        members: this.get('members'),
+        admins: this.get('groupAdmins'),
+      };
+      const groupUpdateMessage = new libsession.Messages.Outgoing.ClosedGroupUpdateMessage(
+        updateParams
       );
+
+      groupUpdate.recipients.forEach(r => {
```
Comment (Mikunj) [inline js/models/conversations.js:1926]: why not
```
libsession.getMessageQueue().sendToGroup(groupUpdateMessage)
```

File: js/models/conversations.js [hunk 12]
```diff
@@ -1930,12 +1992,16 @@
...
-          this.wrapSend(
-            textsecure.messaging.leaveGroup(this.id, groupNumbers, options)
-          )
-        );
+        // FIXME what about public groups?
+        const quitGroup = {
+          timestamp: Date.now(),
+          groupId: this.id,
+        };
+        const quitGroupMessage = new new libsession.Messages.Outgoing.ClosedGroupLeaveMessage(
```
Comment (Mikunj) [inline js/models/conversations.js:2000]:
```suggestion
        const quitGroupMessage = new libsession.Messages.Outgoing.ClosedGroupLeaveMessage(
```

File: libloki/api.js [hunk 13]
```diff
@@ -55,56 +55,12 @@
...
-      );
-    }
-    return new textsecure.protobuf.PairingAuthorisationMessage({
-      requestSignature: new Uint8Array(requestSignature),
-      grantSignature: grantSignature ? new Uint8Array(grantSignature) : null,
-      primaryDevicePubKey,
-      secondaryDevicePubKey,
-    });
-  }
+    const user = libsession.Types.PubKey.from(pubKey);
```
Comment (Mikunj) [inline libloki/api.js:58]:
```suggestion
    const user = new libsession.Types.PubKey(pubKey);
```

File: libloki/api.js [hunk 14]
```diff
@@ -216,57 +172,15 @@
...
-        // eslint-disable-next-line more/no-then
-        ConversationController.getOrCreateAndWait(memberPubKey, 'private').then(
-          () => {
-            const sessionRequestMessage = textsecure.OutgoingMessage.buildSessionRequestMessage(
-              memberPubKey
-            );
-            sessionRequestMessage.sendToNumber(memberPubKey);
-          }
+      if (memberStr !== ourPubKey) {
+        await ConversationController.getOrCreateAndWait(memberStr, 'private');
```
Comment (Mikunj) [inline libloki/api.js:180]: We should be fine not creating a conversation

File: libtextsecure/account_manager.js [hunk 15]
```diff
@@ -545,40 +545,53 @@
...
-      const authorisation = {
-        primaryDevicePubKey,
-        secondaryDevicePubKey: ourPubKey,
-        requestSignature,
-      };
-      await libloki.api.sendPairingAuthorisation(
-        authorisation,
-        primaryDevicePubKey
+
+      const primaryDevice = libsession.Types.PubKey.from(primaryDevicePubKey);
```
Comment (Mikunj) [inline libtextsecure/account_manager.js:549]: `from` can return `undefined` so prefer `new` when you know that it's always going to be a pubkey
```suggestion
      const primaryDevice = new libsession.Types.PubKey(primaryDevicePubKey);
```

File: libtextsecure/message_receiver.js [hunk 16]
```diff
@@ -1086,34 +1086,63 @@ MessageReceiver.prototype.extend({
...
           );
           return;
         }
+        const signedPreKey = {
+          keyId: signedKeyId,
+          publicKey: signedKey,
+          signature,
+        };
+
+        const signedKeyPromise = textsecure.storage.protocol.storeContactSignedPreKey(
```
Comment (Mikunj) [inline libtextsecure/message_receiver.js:1124]: no need to store signed pre key

File: libtextsecure/message_receiver.js [hunk 17]
```diff
@@ -1086,34 +1086,63 @@ MessageReceiver.prototype.extend({
...
+          pubkey,
+          signedPreKey
+        );
+
+        const preKeyObject = {
+          publicKey: preKey,
+          keyId: preKeyId,
+        };
+
+        const preKeyPromise = textsecure.storage.protocol.storeContactPreKey(
```
Comment (Mikunj) [inline libtextsecure/message_receiver.js:1134]: no need to store prekey

File: libtextsecure/outgoing_message.js [hunk 18]
```diff
@@ -216,8 +207,9 @@ OutgoingMessage.prototype = {
...
 
     return (
-      libloki.storage
-        .getAllDevicePubKeysForPrimaryPubKey(primaryPubKey)
+      window.libsession.Protocols.MultiDeviceProtocol.getAllDevices(
+        primaryPubKey
+      )
         // Don't send to ourselves
         .then(devicesPubKeys =>
           devicesPubKeys.filter(pubKey => pubKey !== ourNumber)
```
Comment (Mikunj) [inline libtextsecure/outgoing_message.js:215]:
```suggestion
          devicesPubKeys.filter(pubKey => pubKey.key !== ourNumber)
```

File: ts/receiver/groups.ts [hunk 19]
```diff
@@ -109,12 +111,15 @@ export async function preprocessGroupMessage(
...
     // We have an unknown group, we should request info from the sender
-    window.textsecure.messaging.requestGroupInfo(conversationId, [
-      primarySource,
-    ]);
+    const requestInfo = {
+      timestamp: Date.now(),
+      groupId: conversationId,
+    };
+    const requestInfoMessage = new ClosedGroupRequestInfoMessage(requestInfo);
+    await getMessageQueue().sendToGroup(requestInfoMessage);
```
Comment (Mikunj) [inline ts/receiver/groups.ts:122]: This message needs to be sent to the `sender`

File: ts/session/messages/outgoing/content/data/group/ClosedGroupLeaveMessage.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+  public ttl(): number {
+    return this.getDefaultTTL();
+  }
+
+  protected groupContextType(): SignalService.GroupContext.Type {
+    return SignalService.GroupContext.Type.QUIT;
+  }
+
+  protected dataProto(): SignalService.DataMessage {
+    const messageProto = new SignalService.DataMessage();
```
Comment (Mikunj) [inline ts/session/messages/outgoing/content/data/group/ClosedGroupLeaveMessage.ts:25]: `ClosedGroupMessage` already implements this

File: ts/session/messages/outgoing/content/data/group/ClosedGroupRequestInfoMessage.ts [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  public ttl(): number {
+    return this.getDefaultTTL();
+  }
+
+  protected groupContextType(): SignalService.GroupContext.Type {
+    return SignalService.GroupContext.Type.REQUEST_INFO;
+  }
+
+  protected dataProto(): SignalService.DataMessage {
+    const messageProto = new SignalService.DataMessage();
```
Comment (Mikunj) [inline ts/session/messages/outgoing/content/data/group/ClosedGroupRequestInfoMessage.ts:25]: `ClosedGroupMessage` already implements this

File: ts/session/messages/outgoing/content/sync/SyncMessage.ts [hunk 22]
```diff
@@ -12,5 +13,20 @@ export abstract class SyncMessage extends ContentMessage {
     });
   }
 
+  protected createSyncMessage(): SignalService.SyncMessage {
```
Comment (Mikunj) [inline ts/session/messages/outgoing/content/sync/SyncMessage.ts:16]: wouldn't it be better to stick this code inside `syncProto()`?
The subclasses then can override it and add stuff to it

File: js/models/conversations.js (hunk 11 above)
Comment (Bilb) [reply js/models/conversations.js:1926]: Fixed

File: ts/session/messages/outgoing/content/data/group/ClosedGroupRequestInfoMessage.ts (hunk 21 above)
Comment (Bilb) [reply ts/session/messages/outgoing/content/data/group/ClosedGroupRequestInfoMessage.ts:25]: copy-paste mistake

File: ts/session/messages/outgoing/content/data/group/ClosedGroupLeaveMessage.ts (hunk 20 above)
Comment (Bilb) [reply ts/session/messages/outgoing/content/data/group/ClosedGroupLeaveMessage.ts:25]: copy paste mistake

File: ts/session/messages/outgoing/content/sync/SyncMessage.ts (hunk 22 above)
Comment (Bilb) [reply ts/session/messages/outgoing/content/sync/SyncMessage.ts:16]: Fixed

File: ts/receiver/groups.ts (hunk 19 above)
Comment (Bilb) [reply ts/receiver/groups.ts]: fixed

File: libloki/api.js (hunk 14 above)
Comment (Bilb) [reply libloki/api.js:180]: fixed

File: libtextsecure/account_manager.js (hunk 15 above)
Comment (Bilb) [reply libtextsecure/account_manager.js:549]: fixed

File: libtextsecure/message_receiver.js (hunk 17 above)
Comment (Bilb) [reply libtextsecure/message_receiver.js:1134]: fixed

File: libtextsecure/message_receiver.js (hunk 16 above)
Comment (Bilb) [reply libtextsecure/message_receiver.js:1124]: fixed

File: libloki/api.js (hunk 13 above)
Comment (Bilb) [reply libloki/api.js:58]: fixed

**Pull request merged**